contrib/FreeBSD/sopel

```diff
@@ -0,0 +1,120 @@
+#!/bin/sh
+
+# PROVIDE: sopel
+# REQUIRE: FILESYSTEM NETWORKING usr
+
+. /etc/rc.subr
+
+name="sopel"
+desc="Simple, easy-to-use, open-source IRC utility bot, written in Python"
+rcvar="${name}_enable"
+start_cmd="sopel_start"
+stop_cmd="sopel_stop"
+restart_cmd="sopel_restart"
+status_cmd="sopel_status"
+configure_cmd="sopel_configure"
+extra_commands="configure"
+
+load_rc_config "${name}"
+
+: ${sopel_enable:="NO"}
+: ${sopel_piddir:="/var/run/sopel"}
+: ${sopel_confdir:=/usr/local/etc}
+: ${sopel_flags:=--config-dir "${sopel_confdir}"}
+: ${sopel_program:="/usr/local/bin/sopel"}
+: ${sopel_user:=${name}}
+: ${sopel_interpreter:=/usr/local/bin/python3.9}
+: ${sopel_profiles:=default}
+: ${sopel_prefix:=sopel-}
+: ${sopel_output:=/dev/null}
+
+sopel_start()
+{
+	local profile
+
+	profile="$1"; shift
+
+	echo "Starting sopel profile '${profile}'." && sleep 1
+        /usr/sbin/daemon \
+		-o "${sopel_output}" \
+		-t "${desc}" \
+		-u "${sopel_user}" \
+		${sopel_interpreter} \
+		${sopel_program} start ${sopel_flags} \
+			-c "${sopel_prefix}${profile}" $@
+}
+
+sopel_stop()
+{
+	local pid pidfile profile
+	
+	profile="$1"; shift
+
+	pidfile="${sopel_piddir}/${sopel_prefix}${sopel_prefix}${profile}.pid"
+	if ! [ -f "${pidfile}" ]; then
+		return 1
+	fi
+
+	pid=`cat ${pidfile}`
+
+	echo "Stopping sopel profile '${profile}'."
+        /usr/sbin/daemon \
+		-o "${sopel_output}" \
+		${sopel_interpreter} \
+		${sopel_program} stop ${sopel_flags} \
+			-c "${sopel_prefix}${profile}" $@
+
+	wait_for_pids $pid
+}
+
+sopel_restart()
+{
+	local profile
+	
+	profile="$1"; shift
+
+        run_rc_command stop "${profile}" $@ &&
+        sleep 1 &&
```

Comment:
Necessary?

Reply:
The `sleep 1`?

Reply:
Yeah - will `stop` ever return before sopel has finished stopping? If not, is there a reason for the sleep?

Reply:
Yes, some IRC servers may block an IP with two or more simultaneous connections to prevent flooding. The message `Session limit exceeded` should be printed in the IRC client. I will use `sleep 1` because `daemon(8)` will run sopel in the background without waiting. I know sopel has a flag called --fork, but the daemon is convenient in context because the output is redirected to /dev/null (sopel already prints the output to `/var/log/sopel`, so there is no need to have two outputs).

Reply:
Fair, but I think that belongs in the start function - and it looks like there already is one there, making this one unnecessary, unless I'm misunderstanding `run_rc_command` or something.

Reply:
In the case of `sopel_stop`, there is no `sleep` because the user does not want to wait. In the case of `sopel_start` there is a `sleep` because `sopel_restart` has one. To avoid the above problem, there are 2 seconds to wait, the `sleep` of `sopel_start`, and the `sleep` of `sopel_restart`. I can place `sleep 2` in `sopel_start`, but it may be annoying for the user to wait 2 seconds for each startup profile. The `sleep 1` is unnoticeable.

Reply:
I'm just not clear on why starting multiple bots would need a 1 second delay, but *re*starting would need a 2 second delay. If only 1 second is necessary, the additional delay in restart can be removed. If 2 seconds is always necessary, the start delay should be 2 and the additional delay in restart can still be removed.

Reply:
I understand, I will move the delay from `restart` to `stop`.

2 seconds delay is necessary for fast connections or network delays. It makes more sense to move the delay from `restart` to `stop`.

I understand it is a bit confusing, but the 2 second delay is necessary, and the addition of the same functions makes sense to solve the above problem.

Do we agree with this change?

Reply:
- Why do stops need to be staggered?
- Why do restarts need a larger delay than starts?

Reply:
If I move the delay from `restart` to `stop`, the restart is simply:

```sh
run_rc_command stop "${profile}" $@ &&
run_rc_command start "${profile}" $@
```

The delay is set to `stop`, there is no need to set the delay to `restart` because there is already one.

Why is it necessary to put the delay in `start` and `stop`? Simple: to solve the above problem, it is necessary to delay to 2 seconds. No more, no less.

The delay in `start` is not perceptible. The delay in `stop` is not perceptible.

Yes, i could put the delay in `restart`, but the above problem may occur if the user runs `service sopel stop && service sopel start` instead of `service sopel restart`. I know the user prefers the latter case, but it is possible for it to happen.

Reply:
Sorry, it's not really clear from that - Why do restarts need a larger delay (1s stop + 1s start) than starts (1s)?

Reply:
Due to a slow connection or a network problem that delays a packet before disconnecting it.

Reply:
Normally Sopel sends a `QUIT` message and waits before exiting. Unless this `stop` doesn't wait for sopel to exit, or you know of routine cases I don't, you should be able to assume that if the process exits you're all set.

https://github.com/sopel-irc/sopel/blob/71380df85212747a6c02d8c37c9ffb7836001278/sopel/irc/__init__.py#L635

Reply:
As I mentioned in another comment, the service uses daemon(8) which does not wait for the program to run.

Reply:
Do you agree with the change I mentioned?

Reply:
Starting two:
echo;sleep;fork;echo;sleep;fork;echo;sleep;fork
Restarting two:
stop;wait_for_pids;sleep;echo;sleep;fork;stop;wait_for_pids;sleep;echo;sleep;fork

So for each, to start it simply waits 1 second and then forks, for a total delay between instances of one second.
To stop, it's sending a signal, waiting for Sopel to exit, waiting one second, then in `start` it's waiting another second, then forking the new one. The total delay here is 2 seconds between each instance.

I still don't see why there needs to be twice the delay when restarting vs starting.